File: src/main/java/graveler/GravelerMod.java
package graveler;

import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents;

public class GravelerMod implements ModInitializer {
  public static final String MOD_ID = "graveler";

  @Override
  public void onInitialize() {
    GravelerAttachments.register();

    ServerTickEvents.END_LEVEL_TICK.register(level -> {
      Scheduler scheduler = level.getAttachedOrCreate(GravelerAttachments.SCHEDULER);
      scheduler.tick(level);
    });
  }
}

File: src/main/kotlin/graveler/GravelerMod.kt
package graveler

import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.common.capabilities.CapabilityManager
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext

@Mod(GravelerMod.ModId)
class GravelerMod {
  private fun init(event: FMLCommonSetupEvent) {
    CapabilityManager.INSTANCE.register(
      Scheduler::class.java, SchedulerStorage()) { Scheduler.withDefaults() }

    CapabilityManager.INSTANCE.register(StressMap::class.java, StressMapStorage()) { StressMap(mutableMapOf()) }
  }

  companion object {
    const val ModId = "graveler"
  }

  init {
    val modEventBus = FMLJavaModLoadingContext.get().modEventBus
    modEventBus.addListener { event: FMLCommonSetupEvent -> init(event) }
    MinecraftForge.EVENT_BUS.let {
      it.register(this)
      it.register(PhysicsHandler())
    }
  }
}
